File: elixir/protein-translation/protein_translation.exs
defmodule ProteinTranslation do

  @codon_protein_map %{
    "UGU" => "Cysteine",
    "UGC" => "Cysteine",
    "UUA" => "Leucine",
    "UUG" => "Leucine",
    "AUG" => "Methionine",
    "UUU" => "Phenylalanine",
    "UUC" => "Phenylalanine",
    "UCU" => "Serine",
    "UCC" => "Serine",
    "UCA" => "Serine",
    "UCG" => "Serine",
    "UGG" => "Tryptophan",
    "UAU" => "Tyrosine",
    "UAC" => "Tyrosine",
    "UAA" => "STOP",
    "UAG" => "STOP",
    "UGA" => "STOP"
  }

  @doc """
  Given an RNA string, return a list of proteins specified by codons, in order.
  """
  @spec of_rna(String.t()) :: {atom, list(String.t())}
  def of_rna(rna) do
    # splits rna sequence into list of codons (each codon being a list of 3 graphemes)
    codons = String.splitter(rna, "", trim: true) |> Enum.chunk_every(3)

    proteins = Enum.reduce_while(codons, [],
      fn codon_in_list, acc ->
        codon = Enum.join(codon_in_list) # e.g ["U", "A", "G"] => "UAG"
        protein = @codon_protein_map[codon]

        cond do
          protein != "STOP" ->
            # valid protein
            {:cont, acc ++ [protein]}
          protein == nil ->
            # invalidates rna sequence
            {:halt, [nil]}
          true ->
            # == "STOP"
            {:halt, acc}
        end
      end)

    case Enum.find_index(proteins, &(&1 == nil)) do
      nil -> # index nil so did not find
        {:ok, proteins}
      _ ->
        {:error, "invalid RNA"}
    end
  end

  @doc """
  Given a codon, return the corresponding protein
  """
  @spec of_codon(String.t()) :: {atom, String.t()}
  def of_codon(codon) do
    case @codon_protein_map[codon] do
      nil ->
        {:error, "invalid codon"}
      protein ->
        {:ok, protein}
    end
  end
end

File: erlang/protein-translation/src/protein_translation.erl
-module(protein_translation).

-export([proteins/1]).

-spec proteins(iolist()) -> list(atom()).
proteins(Strand) -> protein_names(Strand, []).

protein_names([], NameList) -> list_names(NameList);
protein_names([P1, P2, P3 | Strand], NameList) -> 
    case name([P1, P2, P3]) of
        stop -> list_names(NameList);
        Name -> protein_names(Strand, [Name | NameList])
    end.

list_names(NameList) -> lists:reverse(NameList).

name("AUG") -> methionine;
name(Name) when Name == "UUU"; Name == "UUC" ->	phenylalanine;
name(Name) when Name == "UUA"; Name == "UUG" -> leucine;
name(Name) when Name == "UCU"; Name == "UCC"; Name == "UCA"; Name == "UCG" -> serine;
name(Name) when Name == "UAU"; Name == "UAC" -> tyrosine;
name(Name) when Name == "UGU"; Name == "UGC" -> cysteine;
name("UGG") -> tryptophan;
name(Name) when Name == "UAA"; Name == "UAG"; Name == "UGA" -> stop.
